File: src/go/wt/internal/worktree/worktree.go
// Package worktree discovers git worktrees.
package worktree

import (
	"encoding/json"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
)

// Info holds the core state for a single worktree.
type Info struct {
	Name      string `json:"name"`
	Path      string `json:"path"`
	Branch    string `json:"branch"`
	IsMain    bool   `json:"is_main"`
	IsCurrent bool   `json:"is_current"`
}

// List discovers all git worktrees via `git worktree list --porcelain`.
func List() ([]Info, error) {
	currentDir, err := os.Getwd()
	if err != nil {
		return nil, err
	}
	currentDir, _ = filepath.EvalSymlinks(currentDir)

	raw, err := exec.Command("git", "worktree", "list", "--porcelain").Output()
	if err != nil {
		return nil, fmt.Errorf("git worktree list: %w", err)
	}

	entries := parseWorktreeList(string(raw))

	// First entry in `git worktree list` is always the main worktree
	mainPath := ""
	if len(entries) > 0 {
		mainPath = entries[0].path
	}

	var infos []Info
	for _, e := range entries {
		info := Info{
			Path:      e.path,
			Branch:    e.branch,
			IsCurrent: currentDir == e.path || strings.HasPrefix(currentDir, e.path+string(filepath.Separator)),
			IsMain:    e.path == mainPath,
		}

		if info.IsMain {
			info.Name = "(main)"
		} else {
			info.Name = filepath.Base(e.path)
		}

		infos = append(infos, info)
	}

	return infos, nil
}

// FindByName returns the worktree matching the given name.
func FindByName(name string) (*Info, error) {
	all, err := List()
	if err != nil {
		return nil, err
	}
	nameLower := strings.ToLower(name)
	for i := range all {
		if strings.ToLower(all[i].Name) == nameLower {
			return &all[i], nil
		}
	}
	return nil, fmt.Errorf("Worktree '%s' not found", name)
}

// Current returns the worktree for the current working directory.
func Current() (*Info, error) {
	all, err := List()
	if err != nil {
		return nil, err
	}
	for i := range all {
		if all[i].IsCurrent {
			return &all[i], nil
		}
	}
	return nil, fmt.Errorf("current directory is not a git worktree")
}

// FormatHuman formats a single worktree as a human-readable line.
func FormatHuman(info *Info) string {
	return fmt.Sprintf("%s  %s", info.Name, info.Branch)
}

// FormatAllHuman formats all worktrees as a human-readable table.
func FormatAllHuman(infos []Info) string {
	repoName := ""
	wtDir := ""
	if len(infos) > 0 {
		for _, info := range infos {
			if info.IsMain {
				repoName = filepath.Base(info.Path)
				wtDir = filepath.Dir(info.Path)
				break
			}
		}
		if repoName == "" {
			repoName = filepath.Base(infos[0].Path)
			wtDir = filepath.Dir(infos[0].Path)
		}
	}

	var sb strings.Builder
	fmt.Fprintf(&sb, "Worktrees for: %s\n", repoName)
	fmt.Fprintf(&sb, "Location: %s\n\n", wtDir)

	for _, info := range infos {
		marker := "  "
		if info.IsCurrent {
			marker = "* "
		}
		fmt.Fprintf(&sb, "%s%-14s %s\n", marker, info.Name, info.Branch)
	}

	fmt.Fprintf(&sb, "\nTotal: %d worktree(s)", len(infos))
	return sb.String()
}

// FormatJSON formats a single worktree as JSON.
func FormatJSON(info *Info) (string, error) {
	data, err := json.MarshalIndent(info, "", "  ")
	if err != nil {
		return "", err
	}
	return string(data), nil
}

// FormatAllJSON formats all worktrees as a JSON array.
func FormatAllJSON(infos []Info) (string, error) {
	data, err := json.MarshalIndent(infos, "", "  ")
	if err != nil {
		return "", err
	}
	return string(data), nil
}

type worktreeEntry struct {
	path   string
	branch string
}

func parseWorktreeList(raw string) []worktreeEntry {
	var entries []worktreeEntry
	var current worktreeEntry

	for _, line := range strings.Split(raw, "\n") {
		if strings.HasPrefix(line, "worktree ") {
			if current.path != "" {
				entries = append(entries, current)
			}
			current = worktreeEntry{path: strings.TrimPrefix(line, "worktree ")}
		} else if strings.HasPrefix(line, "branch refs/heads/") {
			current.branch = strings.TrimPrefix(line, "branch refs/heads/")
		}
	}
	if current.path != "" {
		entries = append(entries, current)
	}
	return entries
}

File: src/fab-rust/src/worktree.rs
use anyhow::{bail, Result};
use serde::Serialize;
use std::fs;
use std::path::{Path, PathBuf};
use std::process::Command;

use crate::status;
use crate::statusfile;

/// Info holds the fab pipeline state for a single worktree.
#[derive(Clone, Debug, Serialize)]
pub struct Info {
    pub name: String,
    pub path: String,
    pub branch: String,
    pub is_main: bool,
    pub is_current: bool,
    pub change: String,
    #[serde(skip_serializing_if = "String::is_empty")]
    pub stage: String,
    #[serde(skip_serializing_if = "String::is_empty")]
    pub state: String,
}

struct WorktreeEntry {
    path: String,
    branch: String,
}

/// Discovers all git worktrees and resolves their fab pipeline state.
pub fn list() -> Result<Vec<Info>> {
    let current_dir = std::env::current_dir()?;
    let current_dir = fs::canonicalize(&current_dir).unwrap_or(current_dir);
    let current_str = current_dir.to_string_lossy().to_string();

    let output = Command::new("git")
        .args(["worktree", "list", "--porcelain"])
        .output()?;
    if !output.status.success() {
        bail!("git worktree list: {}", String::from_utf8_lossy(&output.stderr));
    }

    let raw = String::from_utf8_lossy(&output.stdout).to_string();
    let entries = parse_worktree_list(&raw);

    // First entry is always the main worktree
    let main_path = entries.first().map(|e| e.path.clone()).unwrap_or_default();

    let mut infos = Vec::new();
    for e in &entries {
        let is_current = current_str == e.path
            || current_str.starts_with(&format!("{}/", e.path));

        let mut info = Info {
            path: e.path.clone(),
            branch: e.branch.clone(),
            is_current,
            is_main: e.path == main_path,
            name: if e.path == main_path {
                "(main)".to_string()
            } else {
                Path::new(&e.path)
                    .file_name()
                    .map(|n| n.to_string_lossy().to_string())
                    .unwrap_or_default()
            },
            change: String::new(),
            stage: String::new(),
            state: String::new(),
        };

        resolve_fab_state(&mut info);
        infos.push(info);
    }

    Ok(infos)
}

/// Returns the worktree matching the given name.
pub fn find_by_name(name: &str) -> Result<Info> {
    let all = list()?;
    let name_lower = name.to_lowercase();
    for info in all {
        if info.name.to_lowercase() == name_lower {
            return Ok(info);
        }
    }
    bail!("Worktree '{}' not found", name);
}

/// Returns the worktree for the current working directory.
pub fn current() -> Result<Info> {
    let all = list()?;
    for info in all {
        if info.is_current {
            return Ok(info);
        }
    }
    bail!("current directory is not a git worktree");
}

/// Formats a single worktree as a human-readable line.
pub fn format_human(info: &Info) -> String {
    if !info.stage.is_empty() {
        format!("{}  {}  {}  {}", info.name, info.change, info.stage, info.state)
    } else {
        format!("{}  {}", info.name, info.change)
    }
}

/// Formats all worktrees as a human-readable table.
pub fn format_all_human(infos: &[Info]) -> String {
    let mut repo_name = String::new();
    let mut wt_dir = String::new();

    for info in infos {
        if info.is_main {
            repo_name = Path::new(&info.path)
                .file_name()
                .map(|n| n.to_string_lossy().to_string())
                .unwrap_or_default();
            wt_dir = Path::new(&info.path)
                .parent()
                .map(|p| p.to_string_lossy().to_string())
                .unwrap_or_default();
            break;
        }
    }
    if repo_name.is_empty() && !infos.is_empty() {
        repo_name = Path::new(&infos[0].path)
            .file_name()
            .map(|n| n.to_string_lossy().to_string())
            .unwrap_or_default();
        wt_dir = Path::new(&infos[0].path)
            .parent()
            .map(|p| p.to_string_lossy().to_string())
            .unwrap_or_default();
    }

    let mut sb = String::new();
    sb.push_str(&format!("Worktrees for: {}\n", repo_name));
    sb.push_str(&format!("Location: {}\n\n", wt_dir));

    for info in infos {
        let marker = if info.is_current { "* " } else { "  " };
        if !info.stage.is_empty() {
            sb.push_str(&format!(
                "{}{:<14} {:<42} {:<10} {}\n",
                marker, info.name, info.change, info.stage, info.state
            ));
        } else {
            sb.push_str(&format!("{}{:<14} {}\n", marker, info.name, info.change));
        }
    }

    sb.push_str(&format!("\nTotal: {} worktree(s)", infos.len()));
    sb
}

/// Formats a single worktree as JSON.
pub fn format_json(info: &Info) -> Result<String> {
    Ok(serde_json::to_string_pretty(info)?)
}

/// Formats all worktrees as a JSON array.
pub fn format_all_json(infos: &[Info]) -> Result<String> {
    Ok(serde_json::to_string_pretty(infos)?)
}

fn parse_worktree_list(raw: &str) -> Vec<WorktreeEntry> {
    let mut entries = Vec::new();
    let mut current = WorktreeEntry {
        path: String::new(),
        branch: String::new(),
    };

    for line in raw.lines() {
        if let Some(path) = line.strip_prefix("worktree ") {
            if !current.path.is_empty() {
                entries.push(current);
            }
            current = WorktreeEntry {
                path: path.to_string(),
                branch: String::new(),
            };
        } else if let Some(branch) = line.strip_prefix("branch refs/heads/") {
            current.branch = branch.to_string();
        }
    }
    if !current.path.is_empty() {
        entries.push(current);
    }
    entries
}

fn resolve_fab_state(info: &mut Info) {
    let fab_dir = PathBuf::from(&info.path).join("fab");
    if !fab_dir.exists() {
        info.change = "(no fab)".to_string();
        return;
    }

    let current_file = fab_dir.join("current");
    let data = match fs::read_to_string(&current_file) {
        Ok(d) if !d.trim().is_empty() => d,
        _ => {
            info.change = "(no change)".to_string();
            return;
        }
    };

    let lines: Vec<&str> = data.trim().lines().collect();
    let folder_name = if lines.len() >= 2 {
        lines[1].trim().to_string()
    } else {
        String::new()
    };

    if folder_name.is_empty() {
        info.change = "(no change)".to_string();
        return;
    }

    let status_path = fab_dir
        .join("changes")
        .join(&folder_name)
        .join(".status.yaml");
    if !status_path.exists() {
        info.change = "(stale)".to_string();
        return;
    }

    info.change = folder_name;

    if let Ok(sf) = statusfile::load(&status_path.to_string_lossy()) {
        let (stage, state) = status::display_stage(&sf);
        info.stage = stage;
        info.state = state;
    }
}
